frame/contracts/src/lib.rs

```diff
@@ -949,10 +905,102 @@ pub mod pallet {
 	/// Evicted contracts that await child trie deletion.
 	///
 	/// Child trie deletion is a heavy operation depending on the amount of storage items
-	/// stored in said trie. Therefore this operation is performed lazily in `on_initialize`.
+	/// stored in said trie. Therefore this operation is performed lazily in `on_idle`.
+	#[pallet::storage]
+	pub(crate) type DeletionQueueMap<T: Config> = StorageMap<_, Twox64Concat, u32, DeletedContract>;
+
+	/// A pair of monotonic counters used to index the latest contract marked for deletion
+	/// and the latest deleted contract in DeletionQueueMap.
 	#[pallet::storage]
-	pub(crate) type DeletionQueue<T: Config> =
-		StorageValue<_, BoundedVec<DeletedContract, T::DeletionQueueDepth>, ValueQuery>;
+	pub(crate) type DeletionQueueNonce<T: Config> =
+		StorageValue<_, DeletionQueueNonces, ValueQuery>;
+}
+
+/// DeletionQueueNonces is a pair of monotonic counters used to index contracts marked for deletion,
+/// and the latest deleted contract.
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, Clone, Copy, Default)]
+struct DeletionQueueNonces {
+	/// Monotonic counter used as a key for inserting new deleted contract in the DeletionQueueMap.
+	/// The nonce is incremented after each insertion.
+	insert_nonce: u32,
+	/// The index used to read the next element to be deleted in the DeletionQueueMap.
+	/// The nonce is incremented after each deletion.
+	delete_nonce: u32,
```

Comment:
You should just move those fields into the `DeletionQueue`. You are probably hitting problems with inferred trait bounds of the derive macros. This will help:
```rust
#[scale_info(skip_type_params(T))]
```